README.md
[![Build Status](https://travis-ci.org/sorentwo/proquest.svg?branch=master)](https://travis-ci.org/sorentwo/proquest)
[![npm version](https://badge.fury.io/js/proquest.svg)](http://badge.fury.io/js/proquest)

# Proquest (Promise Request)

Ultra light wrapper around XHR requests that returns Promises for composition.

## Installation

Designed for usage with CJS, namely via [browserify](browserify).

```bash
npm install proquest --save-dev
```

Then simply require it:

```javascript
var Request = require('proquest').Request;
```

## Usage

Proquest takes many of its design cues from [superagent](superagent), but in an
extremely paired back way.

```javascript
// Composing a request

var Request = require('proquest').Request;
var request = new Request('POST', 'https://example.com/things');

request
  .set('Accept', 'application/json')
  .set('Content-Type', 'application/json')
  .set('Authorization', 'token abcdefg1234567')
  .send({ name: 'New Thing', tags: ['a', 'b', 'c'] })
  .end()
  .then(function(response) {
    console.log('Success!', response.body);
  }, function(response) {
    console.log('Failure!', response.status);
  });
```

Often in the context of an application each request will share most headers.
This can be achieved easily with a `partial` request:

```javascript
// Compose a request using method, url, data, or header options.
var partial = Request.partial({
  header: {
    'Authorization' : 'token abc123'
    'Content-Type'  : 'application/json'
    'Accept'        : 'application/json'
  }
});

// Store the partial request for use throughout the application.
MyApp.request = partial;

// Elsewhere, use the request by invoking and completing it.
var request = partial({ method: 'HEAD', url: '/status' });
```

Inject a default error handler into all requests:

```javascript
Request.partial({
  catch: function(response) {
    console.log('failed with a status of: ' + response.status);
  }
});
```

## Notes

Proquest expects a modern XHR api underneath. That means IE8 and below isn't
going to work.

[browserify]: http://browserify.org/
[superagent]: http://visionmedia.github.io/superagent/


CHANGELOG.md
## 0.5.0 - 2016-10-11

* Fixed: Propagate server errors in the 500 range.

## 0.4.0 - 2016-09-08

* Added: Expose an `abort` function on the request object.

## 0.3.0

* Fixed: Swallow errors when parsing JSON. This prevents errors for servers that
  return an HTML error response (404, 500).

## 0.2.0

* Added: Append a `catch` handler on `end` when provided. This allows universal
  error handling.
* Fixed: Properly handle empty xhr responseText
* Fixed: Resolve promises if the xhr status code is between 200 and 399. [Devon
  Blandin]

## v0.1.1 - 2014-09-23

* Added: Handle connection logic within the fulfillment function.
* Fixed: Avoid passing request to response object, just pass the XHR that it
  needs.

## v0.1.0 - 2014-08-12

* Initial release!


test/request_test.js
var Promise  = require('es6-promise').Promise;
var expect   = require('chai').expect;
var sinon    = require('sinon');
var Request  = require('../proquest').Request;
var Response = require('../proquest').Response;

describe('Request', function() {
  var currentRequest;
  var xhr;

  beforeEach(function() {
    xhr = sinon.useFakeXMLHttpRequest();
    global.XMLHttpRequest = xhr;
    global.window = {
      location: {
        hostname: 'example.org'
      }
    };

    xhr.onCreate = function(nextRequest) {
      currentRequest = nextRequest;
    }
  });

  afterEach(function() {
    currentRequest = null;
    xhr.restore();
  });

  describe('.partial', function() {
    it('composes a request with pre-defined attributes', function() {
      var partial = Request.partial({
        method: 'GET',
        header: { 'Accept' : 'application/json' },
        catcher: function(res) {}
      });

      var request = partial({method: 'HEAD', url: '/endpoint'});

      expect(request.method).to.eq('HEAD');
      expect(request.url).to.eq('/endpoint');
      expect(request.header).to.eql({ 'Accept' : 'application/json' });
      expect(request.catcher).to.exist;
    });
  });

  describe('#set', function() {
    it('updates the header object', function() {
      var request = new Request('get', '/');

      request.set('Accept', 'application/json');

      expect(request.header).to.have.key('Accept');
    });
  });

  describe('#catch', function() {
    it('sets a default catcher function', function() {
      var request = new Request('get', '/');
      var pong    = function() {};

      request.catch(pong);

      expect(request.catcher).to.eql(pong);
    });
  });

  describe('#send', function() {
    it('merges the object into data', function() {
      var request = new Request('post', '/');

      request.send({ id: 100, name: 'Hello' });
      expect(request.data).to.eql({ id: 100, name: 'Hello' });

      request.send({ type: 'thing' });
      expect(request.data).to.include({ type: 'thing' });
    });
  });

  describe('#end', function() {
    it('returns a promise', function(done) {
      var request = new Request('POST', '/');
      var promise = request.end();

      expect(promise).to.be.instanceof(Promise);
      done();
    });

    it('performs an XMLHttpRequest', function(done) {
      var method = 'POST';
      var payload = { id: 100, name: 'Hello' };
      var url = '/';
      var request = new Request(method, url);

      request.send(payload);
      request.end();

      expect(currentRequest).to.exist;
      expect(currentRequest.method).to.eq(method);
      expect(JSON.parse(currentRequest.requestBody)).to.eql(payload);
      expect(currentRequest.url).to.eq('/');

      currentRequest.respond(200, {}, '{}');

      done();
    });
  });

  describe('#abort', function() {
    it('cancels a request that has been sent', function(done) {
      var request = new Request('post', '/');
      var promise = request.end();
      expect(request.xhr.readyState).to.eq(1);

      request.abort();
      expect(request.xhr.readyState).to.eq(0);

      promise.catch(function(response) {
        expect(response).to.exist;
        expect(response).to.be.instanceof(Response);
      });

      done();
    });
  });

  describe('#isXDomainRequest', function() {
    it('is a cross domain request when the hostname does not match', function() {
      var request = new Request('get', 'https://example.com/stuff');

      expect(request.isXDomainRequest('localhost:8080')).to.be.true;
      expect(request.isXDomainRequest('example.com')).to.be.false;
    });
  });
});


test/response_test.js
var expect   = require('chai').expect;
var Response = require('../proquest').Response;

describe('Response', function() {
  it('does not try to parse responses without any text', function() {
    var xhr = {
      getAllResponseHeaders: function() { return ''; },
      responseText: '',
      status: 204
    };

    var response = new Response(xhr);

    expect(response.body).to.be.null;
  });

  it('swallows errors during json parsing', function() {
    var xhr = {
      getAllResponseHeaders: function() { return ''; },
      responseText: '<html>Not Found</html>',
      status: 404
    };

    var response = new Response(xhr);

    expect(response.body).to.be.null;
  });

  it('re-throws errors so they can be caught by the request/response consumer', function() {
    var xhr = {
      getAllResponseHeaders: function() { return ''; },
      responseText: '<html>Server Error</html>',
      status: 500
    }

    expect(function() { return new Response(xhr); }).to.throw(Error);
  });
});


lib/response.js
var merge = require('./merge');

var Response = function(xhr) {
  this.xhr    = xhr;
  this.text   = this.xhr.responseText;
  this.header = this.parseHeader(this.xhr.getAllResponseHeaders());
  this.status = this.xhr.status;
  this.body   = null;

  if (this.status < 500) {
    try {
      if (this.text) this.body = JSON.parse(this.text);
    } catch(error) {
      console.error(error);
    }
  } else {
    throw(Error(this.text));
  }
};

merge(Response.prototype, {
  parseHeader: function(string) {
    var lines  = string.split(/\r?\n/);
    var fields = {};
    var index, line, field, value;

    lines.pop(); // trailing CRLF

    for (var i = 0, len = lines.length; i < len; ++i) {
      line  = lines[i];
      index = line.indexOf(':');
      field = line.slice(0, index).toLowerCase();
      value = line.slice(index + 1).trim();
      fields[field] = value;
    }

    return fields;
  }
});

module.exports = Response;


lib/request.js
var merge     = require('./merge');
var ESPromise = require('es6-promise').Promise;
var Response  = require('./response');

var Request = function(method, url) {
  this.method = method;
  this.url    = url;
  this.data   = {};
  this.header = {};
  this.xhr    = new XMLHttpRequest();
};

Request.partial = function(partOpts) {
  return function(options) {
    var merged  = merge(partOpts, options);
    var request = new Request(merged.method, merged.url);

    if (merged.header)  request.header  = merged.header;
    if (merged.data)    request.data    = merged.data;
    if (merged.catcher) request.catcher = merged.catcher;

    return request;
  };
};

merge(Request.prototype, {
  set: function(field, value) {
    this.header[field] = value;

    return this;
  },

  catch: function(func) {
    this.catcher = func;
  },

  send: function(data) {
    merge(this.data, data);

    return this;
  },

  end: function() {
    var request = this;
    var xhr = this.xhr;

    var promise = new ESPromise(function(resolve, reject) {
      xhr.addEventListener('load', function() {
        if (xhr.status >= 200 && xhr.status <= 399) {
          resolve(new Response(xhr));
        } else {
          reject(new Response(xhr));
        }
      });

      xhr.addEventListener('error', function() {
        reject(new Response(xhr));
      });

      xhr.addEventListener('abort', function() {
        reject(new Response(xhr));
      });

      if (request.isXDomainRequest()) {
        xhr.withCredentials = true;
      }

      xhr.open(request.method, request.url, true);

      for (var field in request.header) {
        xhr.setRequestHeader(field, request.header[field]);
      }

      xhr.send(request.serialized());
    });

    if (this.catcher) promise.catch(this.catcher);

    return promise;
  },

  abort: function() {
    this.xhr.abort();
  },

  isXDomainRequest: function(hostname) {
    hostname = hostname || window.location.hostname;
    var hostnameMatch = this.url.match(/http[s]?:\/\/([^\/]*)/);

    return (hostnameMatch && hostnameMatch[1] !== hostname);
  },

  serialized: function() {
    return this.parser()(this.data);
  },

  parser: function() {
    if (this.method !== 'GET' && this.method !== 'HEAD') {
      return JSON.stringify;
    } else {
      return function(value) { return value; };
    }
  }
});

module.exports = Request;
